File: proj.lua/actLua/knight/KnightTitleInfo.lua
require "externCOK2"
require "commonCOK2"

KnightTitleInfo = class("KnightTitleInfo")
KnightTitleInfo.__index = KnightTitleInfo

function KnightTitleInfo:create(params)
    local ct = KnightTitleInfo.new()
    ct:parasDict(params)
    return ct
end

-- KnightTitleInfo.parasDict = function ( params )
--     -- body
-- end

function KnightTitleInfo:parasDict(params)
    -- body

    print("KnightTitleInfo:     params ：" .. tostring(params))

    self.itemId = params["id"]
    self.name = params["name"]
    self.des = params["description"]
    self.icon = params["icon"]
    self.bd_level = params["building_level"]
    self.color = params["color"]
    self.power = params["power"]
    self.unlock = 0

    print("KnightTitleInfo:print" .. tostring(self.itemId))

    self.showDias = {};
    self.values = {};
    self.parasMap = {};
    self.mateUseMap = {};--激活所需材料
    self.mateUnlockMap = {};--解锁所需材料

    if nil ~= params["activate_array"] then
        local vec = string.split(params["activate_array"],"|")
        for k,v in pairs(vec) do
            print(k,v)
            local tmpVec = string.split(v,";")
            local vecCnt = table.getn(tmpVec)
            if (vecCnt >= 2) then
                self.mateUseMap[ tmpVec[1] ] = tmpVec[2]
                print(tmpVec[1],tmpVec[2])
                print("*******************")
            end
        end
    end

    print("*******************  mate end *************")

    if nil ~= params["unlock_array"] then
        local vec = string.split(params["unlock_array"],"|")
        for k,v in pairs(vec) do
            print(k,v)
            local tmpVec = string.split(v,";")
            local vecCnt = table.getn(tmpVec)
            if (vecCnt >= 2) then
                self.mateUnlockMap[ tmpVec[1] ] = tmpVec[2]
                print(tmpVec[1],tmpVec[2])
                print("*******************")
            end
        end
    end

    for i=1,20 do
        local key = "show"..i
        print(tostring(key))
        if nil ~= params[key] then
            self.showDias[i] = params[key]
        else
            break;
        end
    end

    for i=1,20 do
        local key = "num"..i
        print(tostring(key))
        if nil ~= params[key] then
            self.values[i] = params[key]
        else
            break;
        end
    end

    for i=1,20 do
        local key = "para"..i
        print(tostring(key))
        if nil ~= params[key] then
            self.parasMap[i] = string.split(params[key],";")
        else
            break;
        end
    end
end

function KnightTitleInfo:getEffFormatByOrd(order)
    if nil ~= self.parasMap[order] then
        local cnt = table.getn(self.parasMap[order])
        if cnt>0 then
            local eNum = self.parasMap[order][1]
            return CCCommonUtilsForLua:getEffFormat(eNum)
        end
    end
    return "%"
end

function KnightTitleInfo:getEffPMByOrd(order)
    if (nil ~= self.parasMap[order]) then
        local cnt = table.getn(self.parasMap[order])
        if (cnt>0) then
            local eNum = self.parasMap[order][1]
            return CCCommonUtilsForLua:getEffPM(eNum)
        end
    end
    return "+"
end

File: IF/Classes/model/KnightTitleInfo.cpp
//
//  KnightTitleInfo.cpp
//  IF
//
//  Created by 付彬 on 16/1/7.
//
//

#include "KnightTitleInfo.hpp"

KnightTitleInfo::KnightTitleInfo(CCDictionary* dict)
{//<ItemSpec id="260100" level="1" site="0" color="0" name="104301" description="104302" icon="item701" material1="201010;1" material2="201060;1"/>
    this->itemId = dict->valueForKey("id")->intValue();
    this->name = dict->valueForKey("name")->getCString();
    this->des = dict->valueForKey("description")->getCString();
    this->icon = dict->valueForKey("icon")->getCString();
    
    this->bd_level = dict->valueForKey("building_level")->intValue();
    this->color = dict->valueForKey("color")->intValue();
    this->power = dict->valueForKey("power")->intValue();
    
    string useMate = dict->valueForKey("activate_array")->getCString();
    vector<string> tmpUseMateVec;
    CCCommonUtils::splitString(useMate, "|", tmpUseMateVec);
    for (int i=0; i<tmpUseMateVec.size(); i++)
    {
        if ( tmpUseMateVec[i] != "" ) {
            vector<string> tmpVec;
            CCCommonUtils::splitString(tmpUseMateVec[i], ";", tmpVec);
            if (tmpVec.size()>=2) {
                this->mateUseMap[ atoi(tmpVec[0].c_str()) ] = atoi(tmpVec[1].c_str());
            }
        }
    }
    
    string unlockMate = dict->valueForKey("activate_array")->getCString();
    vector<string> tmpUnlockMateVec;
    CCCommonUtils::splitString(unlockMate, "|", tmpUnlockMateVec);
    for (int i=0; i<tmpUnlockMateVec.size(); i++)
    {
        if ( tmpUnlockMateVec[i] != "" ) {
            vector<string> tmpVec;
            CCCommonUtils::splitString(tmpUnlockMateVec[i], ";", tmpVec);
            if (tmpVec.size()>=2) {
                this->mateUnlockMap[ atoi(tmpVec[0].c_str()) ] = atoi(tmpVec[1].c_str());
            }
        }
    }
    
    for (int i=1; i<=20; i++) {
        string key = "show";
        key = key+CC_ITOA(i);
        if (dict->objectForKey(key)) {
            this->showDias[i] = dict->valueForKey(key)->getCString();
        }else{
            break;
        }
    }
    
    for (int j=1; j<=20; j++) {
        string key = "num";
        key = key+CC_ITOA(j);
        if (dict->objectForKey(key)) {
            this->values[j] = dict->valueForKey(key)->floatValue();
        }else{
            break;
        }
    }
    
    for (int m=1; m<=20; m++) {
        string key = "para";
        key = key+CC_ITOA(m);
        if (dict->objectForKey(key)) {
            string tmpStr = dict->valueForKey(key)->getCString();
            vector<string> tmpVec;
            CCCommonUtils::splitString(tmpStr, "|", tmpVec);
            for (int n=0; n<tmpVec.size(); n++) {
                this->parasMap[m].push_back(atoi(tmpVec[n].c_str()));
            }
        }else{
            break;
        }
    }
}

string KnightTitleInfo::getEffFormatByOrd(int order)
{
    if (parasMap.find(order) != parasMap.end()) {
        if (parasMap[order].size()>0) {
            int eNum = parasMap[order][0];
            return CCCommonUtils::getEffFormat( CC_ITOA(eNum) );
        }
    }
    return "%";
}

string KnightTitleInfo::getEffPMByOrd(int order)
{
    if (parasMap.find(order) != parasMap.end()) {
        if (parasMap[order].size()>0) {
            int eNum = parasMap[order][0];
            return CCCommonUtils::getEffPM( CC_ITOA(eNum) );
        }
    }
    return "+";
}